src/formats/roam-json.ts

```diff
@@ -0,0 +1,415 @@
+		this.progress = progress;
+		let { files } = this;
+		if (files.length === 0) {
+			new Notice('Please pick at least one file to import.');
+			return;
+		}
+
+		let outputFolder = await this.getOutputFolder();
+		if (!outputFolder) {
+			new Notice('Please select a location to export to.');
+			return;
+		}
+
+		for (let file of files) {
+			if (progress.isCancelled()) {
+				return;
+			}
+
+			const graphName = sanitizeFileName(file.basename);
+			const graphFolder = `${outputFolder.path}/${graphName}`;
+			const attachmentsFolder = `${outputFolder.path}/${graphName}/Attachments`;
+
+			// create the base graph folders
+			await this.createFolders(graphFolder);
+			await this.createFolders(attachmentsFolder);
+
+			// read the graph
+			const data = await file.readText();
+			const allPages = JSON.parse(data) as RoamPage[];
+
+			// PRE-PROCESS: map the blocks for easy lookup //
+			const [blockLocations, toPostProcess] = this.preprocess(allPages);
+
+			const markdownPages: Map<string, string> = new Map();
+			for (let index in allPages) {
+				const pageData = allPages[index];
+
+				let pageName = convertDateString(sanitizeFileNameKeepPath(pageData.title), this.userDNPFormat).trim();
+				if (pageName === '') {
+					progress.reportFailed(pageData.uid, 'Title is empty');
+					console.error('Cannot import data with an empty title', pageData);
+					continue;
+				}
+				const filename = `${graphFolder}/${pageName}.md`;
+
+				const markdownOutput = await this.jsonToMarkdown(graphFolder, attachmentsFolder, pageData);
+				markdownPages.set(filename, markdownOutput);
+			}
+
+			// POST-PROCESS: fix block refs //
+			for (const [_, callingBlock] of toPostProcess.entries()) {
+				const callingBlockStringScrubbed = await this.roamMarkupScrubber(graphFolder, attachmentsFolder, callingBlock.blockString, true);
+				const newCallingBlockReferences = await this.extractAndProcessBlockReferences(markdownPages, blockLocations, graphFolder, callingBlockStringScrubbed);
+
+				const callingBlockFilePath = `${graphFolder}/${callingBlock.pageName}.md`;
+				const callingBlockMarkdown = markdownPages.get(callingBlockFilePath);
+				if (callingBlockMarkdown) {
+					let lines = callingBlockMarkdown.split('\n');
+
+					let index = lines.findIndex((item: string) => item.contains('* ' + callingBlockStringScrubbed));
+					if (index !== -1) {
+						lines[index] = lines[index].replace(callingBlockStringScrubbed, newCallingBlockReferences);
+					}
+
+					markdownPages.set(callingBlockFilePath, lines.join('\n'));
+				}
+			}
+
+			// WRITE-PROCESS: create the actual pages //
+			const { vault } = this;
+			const totalCount = markdownPages.size;
+			let index = 1;
+			for (const [filename, markdownOutput] of markdownPages.entries()) {
+				if (progress.isCancelled()) {
+					return;
+				}
+
+				try {
+					//create folders for nested pages [[some/nested/subfolder/page]]
+					await this.createFolders(path.dirname(filename));
```

Comment:
Let's not use path.dirname since it doesn't work on mobile. Instead there's a parsing util function in filesystem.ts.